File: internal/physics/collisions.go
package physics

import (
	"math"

	"github.com/1001bit/OnlineCanvasGames/internal/mathobjects"
)

func CollideKinematicWithStatic(kinematic Kinematic, static Physical, dtMs float64) {
	if !static.CanCollide() {
		return
	}

	// TODO: skip if kinematic->solid direction == kinematic.collisionDir

	// "path", that rect is going to pass
	futureKinematic := kinematic.GetRect()

	// Getting final value of velocity, which will be added to kinematicRect
	finalVel := kinematic.GetVelocity().Scale(dtMs)

	// Vertical
	if finalVel.Y > 0 {
		// down
		futureKinematic.Size.Y += finalVel.Y

		if futureKinematic.Intersects(static.GetRect()) {
			newPosY := static.GetPosition().Y - kinematic.GetSize().Y

			kinematic.SetPosition(kinematic.GetPosition().X, newPosY)
			kinematic.SetVelocity(kinematic.GetVelocity().X, 0)

			kinematic.SetCollisionDir(mathobjects.Down)
		}
	} else if finalVel.Y < 0 {
		// up
		futureKinematic.Size.Y += math.Abs(finalVel.Y)
		futureKinematic.Position.Y -= math.Abs(finalVel.Y)

		if futureKinematic.Intersects(static.GetRect()) {
			newPosY := static.GetPosition().Y + static.GetSize().Y

			kinematic.SetPosition(kinematic.GetPosition().X, newPosY)
			kinematic.SetVelocity(kinematic.GetVelocity().X, 0)

			kinematic.SetCollisionDir(mathobjects.Up)
		}
	}

	futureKinematic = kinematic.GetRect()

	// Horizontal
	if finalVel.X > 0 {
		// Right
		futureKinematic.Size.X += finalVel.X

		if futureKinematic.Intersects(static.GetRect()) {
			newPosX := static.GetPosition().X - kinematic.GetSize().X

			kinematic.SetPosition(newPosX, kinematic.GetPosition().Y)
			kinematic.SetVelocity(0, kinematic.GetVelocity().Y)

			kinematic.SetCollisionDir(mathobjects.Right)
		}
	} else if finalVel.X < 0 {
		// Left
		futureKinematic.Size.X += math.Abs(finalVel.X)
		futureKinematic.Position.X -= math.Abs(finalVel.X)

		if futureKinematic.Intersects(static.GetRect()) {
			newPosX := static.GetPosition().X + static.GetSize().X

			kinematic.SetPosition(newPosX, kinematic.GetPosition().Y)
			kinematic.SetVelocity(0, kinematic.GetVelocity().Y)

			kinematic.SetCollisionDir(mathobjects.Left)
		}
	}
}

File: web/typescript/gameassets/physics/collisions.ts
function collideKinematicWithStatic(kinematicRect: KinematicRect, staticRect: PhysicalRect, dt: number){
    if(!staticRect.doApplyCollisions){
        return
    }
    kinematicRect.setCollisionDir(Direction.None)

    let futureKinematic = new Rect(kinematicRect);
    futureKinematic.setPosition(kinematicRect.targetPosition.x, kinematicRect.targetPosition.y)

    const velX = kinematicRect.velocity.x * dt
    const velY = kinematicRect.velocity.y * dt

    if(velY > 0) {
        // down
        futureKinematic.size.y += velY

        if(futureKinematic.intersects(staticRect)){
            kinematicRect.setTargetPos(kinematicRect.targetPosition.x, staticRect.position.y - kinematicRect.size.y, true)
            kinematicRect.velocity.y = 0

            kinematicRect.setCollisionDir(Direction.Down)
        }

    } else if(velY < 0) {
        // up
        futureKinematic.size.y += Math.abs(velY)
        futureKinematic.position.y -= Math.abs(velY)

        if(futureKinematic.intersects(staticRect)){
            kinematicRect.setTargetPos(kinematicRect.targetPosition.x, staticRect.position.y + staticRect.size.y, true)
            kinematicRect.velocity.y = 0

            kinematicRect.setCollisionDir(Direction.Up)
        }
    }

    futureKinematic = new Rect(kinematicRect);
    futureKinematic.setPosition(kinematicRect.targetPosition.x, kinematicRect.targetPosition.y)

    if(velX > 0) {
        // right
        futureKinematic.size.x += velX

        if(futureKinematic.intersects(staticRect)){
            kinematicRect.setTargetPos(staticRect.position.x - kinematicRect.size.x, kinematicRect.targetPosition.y, true)
            kinematicRect.velocity.x = 0

            kinematicRect.setCollisionDir(Direction.Right)
        }

    } else if(velX < 0) {
        // left
        futureKinematic.size.x += Math.abs(velX)
        futureKinematic.position.x -= Math.abs(velX)

        if(futureKinematic.intersects(staticRect)){
            kinematicRect.setTargetPos(staticRect.position.x + staticRect.size.x, kinematicRect.targetPosition.y, true)
            kinematicRect.velocity.x = 0

            kinematicRect.setCollisionDir(Direction.Left)
        }
    }
}